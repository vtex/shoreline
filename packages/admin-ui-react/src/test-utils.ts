import type { ReactElement } from 'react'
import type { RenderOptions } from '@testing-library/react'
import { render as baseRender } from '@testing-library/react'
import { createSystem } from './system'

const [ThemeProvider] = createSystem()

function render(ui: ReactElement, options?: Omit<RenderOptions, 'queries'>) {
  return baseRender(ui, { wrapper: ThemeProvider, ...options })
}

export { render }
