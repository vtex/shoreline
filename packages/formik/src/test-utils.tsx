import '@testing-library/jest-dom/extend-expect'
import React, { useRef } from 'react'
import { axe, toHaveNoViolations } from 'jest-axe'
import type { RenderOptions } from '@testing-library/react'
import {
  render as baseRender,
  waitFor,
  screen,
  fireEvent,
} from '@testing-library/react'
import { createSystem } from '@vtex/admin-ui-react'
import type { ReactElement } from 'react'
import { renderHook, act } from '@testing-library/react-hooks'

expect.extend(toHaveNoViolations)

const [ThemeProvider] = createSystem()

function render(ui: ReactElement, options?: Omit<RenderOptions, 'queries'>) {
  return baseRender(ui, { wrapper: ThemeProvider, ...options })
}

/**
 * HOC to wrap a component with its state hook
 * @example
 * // Cmp has the useCmp state hook
 * const StatefulCpm = withState(Cmp, () => useCmpState({ isSelected: false }))
 */
function withState(Component: any, useStateHook: () => any) {
  function StatefulComponent(props: any) {
    // Avoids an invalid hook call
    const ref = useRef(useStateHook())

    return <Component state={ref.current} {...props} />
  }

  StatefulComponent.displayName = Component.displayName

  return StatefulComponent
}

function jestMatchMedia() {
  /**
   * ðŸš§ Workaround for window.match media
   * @see https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
   */
  return Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  })
}

export {
  render,
  withState,
  axe,
  jestMatchMedia,
  renderHook,
  act,
  waitFor,
  screen,
  fireEvent,
}
