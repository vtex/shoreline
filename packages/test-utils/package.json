{
  "name": "@vtex/shoreline-test-utils",
  "version": "1.0.32",
  "description": "Shoreline test utilities",
  "main": "./dist/index.mjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.mts",
  "repository": "vtex/shoreline",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "files": [
    "dist"
  ],
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "types": "./dist/index.d.mts"
    },
    "./setup": {
      "import": "./dist/setup.mjs",
      "types": "./dist/setup.d.mts"
    }
  },
  "engines": {
    "node": ">=20"
  },
  "scripts": {
    "prebuild": "rm -rf dist",
    "dev": "tsup --watch",
    "build": "npm run prebuild && tsup"
  },
  "peerDependencies": {
    "vitest": "0.x"
  },
  "devDependencies": {
    "vitest": "0.34.6"
  },
  "dependencies": {
    "@testing-library/dom": "9.3.4",
    "@testing-library/jest-dom": "6.4.6",
    "@testing-library/react": "14.3.1",
    "@testing-library/user-event": "14.5.2"
  }
}
