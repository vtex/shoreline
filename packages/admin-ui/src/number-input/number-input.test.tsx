import React from 'react'

import { render, jestMatchMedia } from '../test-utils'

import type { NumberInputProps, NumberInputValue } from './index'
import { NumberInput } from './index'

const ControlledNumberInput = (props: NumberInputProps) => {
  const [value, setValue] = React.useState<NumberInputValue>(8)
  const { min = Infinity, max = -Infinity } = props

  return (
    <NumberInput
      id="number-input"
      value={value}
      onChange={(newValue: NumberInputValue) => setValue(newValue)}
      error={value > max || value < min}
      {...props}
    />
  )
}

describe('Number Input', () => {
  beforeEach(jestMatchMedia)

  it('should match snapshot', () => {
    const { asFragment } = render(
      <ControlledNumberInput
        label="Label"
        min={-5}
        max={5}
        helpText="Help Text"
        errorText="Error text"
        error
        disabled
        optional
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should load controlled value', () => {
    const { getByTestId } = render(
      <ControlledNumberInput data-testid="number-input" />
    )

    const numberInput = getByTestId('number-input')

    expect(numberInput).toHaveDisplayValue('8')
  })

  it('should load with a default value', () => {
    const { getByTestId } = render(
      <NumberInput
        id="number-input"
        data-testid="number-input"
        label="Label"
        value="5"
      />
    )

    const numberInput = getByTestId('number-input')

    expect(numberInput).toHaveDisplayValue('5')
  })

  it('should show invalid number error', () => {
    const min = -5
    const max = 5

    const { getByText } = render(
      <ControlledNumberInput
        data-testid="number-input"
        label="Label"
        min={min}
        max={max}
        errorText="Invalid number input"
      />
    )

    const errorTextElement = getByText('Invalid number input')

    expect(errorTextElement).toBeVisible()
  })

  it('should show help text', () => {
    const { getByText } = render(
      <NumberInput
        id="number-input"
        data-testid="number-input"
        label="Label"
        helpText="Set its value"
      />
    )

    const helpTextElement = getByText('Set its value')

    expect(helpTextElement).toBeVisible()
  })
})
