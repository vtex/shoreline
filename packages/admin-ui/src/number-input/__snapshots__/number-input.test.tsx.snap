// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number Input should match snapshot 1`] = `
<DocumentFragment>
  .emotion-14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-pack: unset;
  -ms-flex-pack: unset;
  -webkit-justify-content: unset;
  justify-content: unset;
}

.emotion-14 >*:not(:first-child) {
  margin-top: 0.25rem;
}

.emotion-0 {
  color: #545454;
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 400;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.002em;
}

.emotion-10 {
  height: 2.75rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 400;
  font-size: 0.875rem;
  line-height: 1.4285714286;
  letter-spacing: 0.002em;
  border: 1px solid #747474;
  border-radius: 4px;
  background: #E6E6E6;
  cursor: text;
  color: #747474;
}

.emotion-10:hover {
  border: 1px solid #747474;
}

.emotion-10:focus-within {
  border: 1px solid #782020;
  box-shadow: 0rem 0rem 0rem 0.0625rem #FBF2F2,0rem 0rem 0rem 0.1875rem #F49494;
}

.emotion-10:not(:hover):not(:focus-within)>button {
  display: none;
}

.emotion-1 {
  height: 2.75rem;
  padding: 0.75rem 1rem;
  width: 100%;
  border-radius: 4px;
  background: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -moz-appearance: textfield;
}

.emotion-1:focus {
  outline: none;
}

.emotion-1::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0rem;
}

.emotion-1::-webkit-input-placeholder {
  font-style: italic;
}

.emotion-1::-moz-placeholder {
  font-style: italic;
}

.emotion-1:-ms-input-placeholder {
  font-style: italic;
}

.emotion-1::placeholder {
  font-style: italic;
}

.emotion-1:disabled~button {
  display: none;
}

.emotion-5 {
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 500;
  font-size: 0.875rem;
  line-height: 1.1428571429;
  letter-spacing: 0.01em;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  position: relative;
  color: #2953B2;
  background: transparent;
  padding: 0.5rem 0.75rem;
  height: 2.25rem;
  margin-top: 0.25rem;
  margin-bottom: 0.25rem;
}

.emotion-5:hover {
  color: #20408C;
  background: #F1F4FC;
}

.emotion-5:active {
  color: #172F66;
  background: #DDE5F7;
}

.emotion-5:disabled {
  background: transparent;
  color: #747474;
}

.emotion-5:focus {
  outline: none;
  box-shadow: 0rem 0rem 0rem 0.0625rem #F1F4FC,0rem 0rem 0rem 0.1875rem #96B2F2;
}

.emotion-5:focus:not([data-focus-visible-added]) {
  outline: none;
  box-shadow: none;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 500;
  font-size: 0.875rem;
  line-height: 1.1428571429;
  letter-spacing: 0.01em;
  visibility: visible;
}

.emotion-2 {
  height: 1.25rem;
  width: 1.25rem;
  min-height: 1.25rem;
  min-width: 1.25rem;
}

.emotion-9 {
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 500;
  font-size: 0.875rem;
  line-height: 1.1428571429;
  letter-spacing: 0.01em;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  position: relative;
  color: #2953B2;
  background: transparent;
  padding: 0.5rem 0.75rem;
  height: 2.25rem;
  margin-right: 0.25rem;
  margin-top: 0.25rem;
  margin-bottom: 0.25rem;
}

.emotion-9:hover {
  color: #20408C;
  background: #F1F4FC;
}

.emotion-9:active {
  color: #172F66;
  background: #DDE5F7;
}

.emotion-9:disabled {
  background: transparent;
  color: #747474;
}

.emotion-9:focus {
  outline: none;
  box-shadow: 0rem 0rem 0rem 0.0625rem #F1F4FC,0rem 0rem 0rem 0.1875rem #96B2F2;
}

.emotion-9:focus:not([data-focus-visible-added]) {
  outline: none;
  box-shadow: none;
}

.emotion-13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-pack: unset;
  -ms-flex-pack: unset;
  -webkit-justify-content: unset;
  justify-content: unset;
}

.emotion-13 >*:not(:first-child) {
  margin-top: 0rem;
}

.emotion-11 {
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 400;
  font-size: 0.75rem;
  line-height: 1.3333333333;
  letter-spacing: -0.004em;
  color: #545454;
}

.emotion-12 {
  font-family: "VTEX Trust",-apple-system,system-ui,BlinkMacSystemFont,sans-serif;
  font-variation-settings: 'WGHT' 400;
  font-size: 0.75rem;
  line-height: 1.3333333333;
  letter-spacing: -0.004em;
  color: #CC3D3D;
}

<div
    class="emotion-14"
  >
    <label
      class="emotion-0"
      for="number-input"
    >
      Label (optional)
    </label>
    <div
      class="emotion-10"
    >
      <input
        class="emotion-1"
        disabled=""
        id="number-input"
        max="5"
        min="-5"
        placeholder=""
        step="1"
        type="number"
        value="8"
      />
      <button
        aria-label="Decrement"
        class="emotion-5"
        type="button"
      >
        <div
          class="emotion-4"
        >
          <div
            class="emotion-3"
          >
            <svg
              class="emotion-2"
              fill="currentColor"
              height="1.25rem"
              viewBox="0 0 256 256"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="16rem"
                width="16rem"
              />
              <polyline
                fill="none"
                points="208 96 128 176 48 96"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="18"
              />
            </svg>
          </div>
        </div>
      </button>
      <button
        aria-disabled="true"
        aria-label="Increment"
        class="emotion-9"
        disabled=""
        style="pointer-events: none;"
        type="button"
      >
        <div
          class="emotion-4"
        >
          <div
            class="emotion-3"
          >
            <svg
              class="emotion-2"
              fill="currentColor"
              height="1.25rem"
              viewBox="0 0 256 256"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="16rem"
                width="16rem"
              />
              <polyline
                fill="none"
                points="48 160 128 80 208 160"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="18"
              />
            </svg>
          </div>
        </div>
      </button>
    </div>
    <div
      class="emotion-13"
    >
      <span
        class="emotion-11"
      >
        Help Text
      </span>
      <span
        class="emotion-12"
        role="alert"
      >
        Error text
      </span>
    </div>
  </div>
</DocumentFragment>
`;
